### ISSUE TITLE
Not validating source of callback

### ISSUE THREAD

**Issue by ha1331** [liamcurry/passport-steam#35]
There currently is an exploit that uses the fact that surprisingly many applications are not validating the source of the identity. It's pretty easy to fake a request to the url `/auth/steam/return` in the example you have provided. I'm not claiming this is an issue in passport.js or in passport-steam package. It how ever has been a problem for multiple sites not implementing it properly and I know bug bounties have been payed in the range of tens of thousands of dollars. 

As a fix I would suggest you at least update the example to contain a way to verify the source of the provided identity. Better way might be preventing fetching the identity from other servers than steam, since this passport.js strategy is specific to steam. 

Not currently using passport.js for anything and am not in any way an expert, so I can't tell you how to do it. I tested one example I was provided and I was able to "fix" the exploit by checking that `identifier` in `function(identifier, profile, done) {`  has http://steamcommunity.com or https://steamcommunity.com instead of something else. I don't know if `identity` can be faked or not, but it did at least block the malicious url I was given as an example. How ever, as I said, I don't really know the inner workings of passport.js or openid so I can't say if this is enough to detect malicious user.

**Comment by mnzt** [comment]
You raise a very good point, there is quite a high potential for mitm attack for openid and steam.

I don't see any reason the package cannot verify the source on the return payload - I'll see what I can muster up after work

**Comment by breatoz** [comment]
Something like the following would prevent this.

`app.get('/auth/steam/callback',function(req,res,next) {
  if (req.query['openid.claimed_id'].substr(0,26) != "http://steamcommunity.com/"
  || req.query['openid.identity'].substr(0,26) != "http://steamcommunity.com/"
  || req.query['openid.op_endpoint'].substr(0,27) != "https://steamcommunity.com/") {
        res.redirect('/auth/steam');
        return false;
   }

        passport.authenticate('steam', { failureRedirect: '/' })(req,res, next);

}, function(req, res) { req.redirect('/');  });
`

**Comment by breatoz** [comment]
Can you delete this post for now, we're working on patching it and don't want others exploiting sites.

Thx.

**Comment by mnzt** [comment]
@breatoz No, we will not delete this. I am working on a patch which ought to clear this up shortly

**Comment by welps** [comment]
@mnzt Somewhat related to this is #24 -- we're blocked by upstream still utilizing node-openid 1.x.x.   I've been meaning to open a PR to investigate and get passport-openid to using node-openid 2.x.x but literally have had no time since starting new job.

**Comment by mnzt** [comment]
@welps I've had to open up my own fork to fix (#34). We can probably move forward on this?

**Comment by welps** [comment]
@mnzt  oh, I see. I could've sworn that node-openid 2.x.x has breaking changes that passport-openid has to be updated for. Is that not the case?

**Comment by mnzt** [comment]
@welps Not from what I can see, I tested the example running on node-openid 2.x.x and it all seemed fine

**Comment by mnzt** [comment]
@ha1331 just to note, as @breatoz posted before deleting, you can temporarily verify via

```
if (req.query['openid.claimed_id'].substr(0,26) != "http://steamcommunity.com/"
|| req.query['openid.identity'].substr(0,26) != "http://steamcommunity.com/"
|| req.query['openid.op_endpoint'].substr(0,27) != "https://steamcommunity.com/") {
res.redirect('/auth/steam');
return false;
}
```

You may want to be aware of SSL too.

**Comment by kondax** [comment]
I don't believe this is a valid issue. 

I've been attempting many ways to bypass what you seem as no validation in the identified, yet all have failed.

Node-OpenID checks if op_endpoint from the assertion URL is the same as the  endpoint from the provider URL, and if it isn't then it fails.

Without this, you cannot change the other values of the domain. Without the secret, you are not able to generate a valid signature. Without a signature that matches the parameters, the verification call fails.

Unless you can find the secret before verifying the assertion URL, you can't modify any IDs or other parameters. This would, like you said, be an issue in Node-OpenID and not passport-steam since the secret is never released.

**Comment by ha1331** [comment]
@kondax I believe this is as valid issue as issue can be. Seeing how multiple sites have been extorted to pay for a fix, bug bounties has been payed in the tens of thousands and this "bug" or rather limiter or not as secure as possible default implementation has been used to make tens of thousands by logging in as any user you know the steam id for. 

I really have difficulties understanding how you would think it's not an issue anyone can just "pretend to be steam", point a browser to your callback url and tell it to verify the identity from iamsteam.exploiters.com. Can you present any scenario where steam passport strategy would need to verify the users identity from other places than steam?

**Comment by kondax** [comment]
@ha1331 As I said, OpenID already checks if the callback url matches the strategy's domain, hence why I said it isn't valid.

**Comment by ha1331** [comment]
@kondax Well if you say so. Tho the reality of me actually copy pasting an url where I only changed the domain/callback location and steam id  to be able to login to multiple sites, including the example implementation from this repo as any user that existed and you saying it doesnt work are not in sync. Don't know how valid your claim is, but I know this works, since I have tried it. Also I know for a fact this has been used to steal skins. This has been used to extort actual dollars for a fix.Are you saying they just claimed it works and these site owners just didn't check, just payed to a fix for a problem that doesn't exist and imagined they experienced losses in excess of tens of thousands?

**Comment by kondax** [comment]
It's strange, since I tried the same as you without success. I attempted to change the op_endpoint to a fake site that returns the is valid response, yet OpenID checks if the domain is the same. I'm unsure of how you achieved it. 

Sorry, my mistake.

**Comment by ha1331** [comment]
We are approaching the limit of my understanding of how openid works, but the way it works is, fakesteam.exploit.com verifies that the user is valid user for fakesteam.exploit.com and that seems to be enough. Problem here is allowing these verifications to other domains than steamcommunity.com. Does that make sense?

**Comment by welps** [comment]
@ha1331 do you mind trying to reproduce this on the latest version of passport-steam? 

I believe @mnzt updated this package to use a fork of passport-openid where its dependency, node-openid, has been updated to 2.x.x. That version of node-openid addresses the concern brought up in #24

**Comment by kondax** [comment]
@ha1331 then I believe it has been addressed, since with the latest version of passport-steam the exploit seems not to be present.

**Comment by ha1331** [comment]
@welps I tested it, after wiping `node_modules` and doing `npm install` I'm happy to say the "exploit" as it was, does not work anymore. 

@kondax I apologize if my responses came out strong, did not know work had been done to fix this, seeing how the issue was still open and all that.

**Comment by welps** [comment]
@ha1331 Great to hear. Thanks for reporting the issue.

@mnzt Thanks for the update!

Yeah, sorry for the miscommunication guys. We'll try to keep these tickets up-to-date.

**Comment by mnzt** [comment]
Great work on the :bug: squishing, on behalf of the users of this package, I highly appreciate it

**Comment by Pepsai** [comment]
Anyone want to try out to see if this can be exploited on my website? Or teach me on how this can be done, im still very unsure of this issue.

**Comment by bartlomiej-korpus** [comment]
This bug is NOT fixed. Using passport-steam@1.0.2 and openid@2.0.1 I can successfully trick example app (examples/signon) to think I am somebody else.

**Comment by mnzt** [comment]
Can you provide evidence and a steps to reproduce? In am email would be
preferable (you can find my address on my profile)

**Comment by welps** [comment]
cc me too please

**Comment by bartlomiej-korpus** [comment]
Confirmed the issue with @mnzt .

**Comment by kondax** [comment]
Is this using the same 'exploit', or is there another way around the new prevention checks?

**Comment by mnzt** [comment]
Hi, I've been doing some testing with @bartlomiej-korpus and I can confirm this is a vulnerable issue.

Unfortunately: this permits to the openid spec, so we're having to come up with our own fix, thankfully @bartlomiej-korpus is willing to devote his time into a fix.
Fortunately: It's not so easy to do, however it is quite trivial.

**Comment by bartlomiej-korpus** [comment]
@kondax it is the same, in was never fixed in the first place.

**Comment by kondax** [comment]
Gotcha. If you need any help, I'm available.

**Comment by andrewda** [comment]
I can definitely help out, too, and I'll start doing some testing on the best way to avoid this issue. This certainly needs to be fixed ASAP.

**Comment by Pepsai** [comment]
Hopefully this will be fixed soon, thanks a lot for all the hard work you are doing! :+1:

**Comment by bartlomiej-korpus** [comment]
I think the suitable way is to check if the claimedIdentifier after successfull assertion fits the regex like `^http:\/\/steamcommunity.com\/openid\/id\/\d+$/`
What do you guys think?

**Comment by welps** [comment]
@bartlomiej-korpus I didn't get to observe the exploit, but if you exploited the way I think you did, then @mnzt 's comment for a fix works too, no? https://github.com/liamcurry/passport-steam/issues/35#issuecomment-217896085

**Comment by breatoz** [comment]
Yes; just double check that I had those variable names right; and its assertive ;p

Sorry I've been swamped and haven't participated more

**Comment by bartlomiej-korpus** [comment]
Probably yes, but thats not the most elegant way to protect

**Comment by mnzt** [comment]
@welps our main concern with that is the ability to use a subdomain, for example `steamcommunity.com.github.com`, allong with the usage of nullbytes and other such exploits

**Comment by bartlomiej-korpus** [comment]
I have mailed node-openid author, person experienced with openid protocol to consult the fix with regexes, hope he agrees that it is enough.

**Comment by welps** [comment]
@mnzt gotcha, seems like regex should work out too then

**Comment by kondax** [comment]
I can get the 'exploit' to work with the latest version on my local server (changing all addresses to http://steamcommunity.com.<xxx>:3000), but I am not able to reproduce the error on an older version on my development server.

**Comment by mnzt** [comment]
@kondax can you please provide version numbers

**Comment by kondax** [comment]
@mnzt Hang on, I'll grab them in a bit. I'll edit this with them.

**Comment by Pepsai** [comment]
Any news on this?

**Comment by waylaidwanderer** [comment]
I had a discussion with one of the exploiters a few weeks ago and he told me that my site which uses https://github.com/invisnik/laravel-steam-auth is not vulnerable. Although it is PHP, perhaps you can take a look at their [`validate()`](https://github.com/invisnik/laravel-steam-auth/blob/master/src/SteamAuth.php#L50) implementation.

**Comment by kondax** [comment]
@waylaidwanderer 

That library isn't vulnerable since it doesn't call the URL found in the OpenID return parameters but rather a static steamcommunity URL with the OpenID parameters plugged in.

Using the regex @bartlomiej-korpus provided would be enough, but it's also possible to implement a solution that laravel-steam-auth uses.

**Comment by waylaidwanderer** [comment]
>it doesn't call the URL found in the OpenID return parameters but rather a static steamcommunity URL with the OpenID parameters plugged in

Yep, I'm aware. I think this would be the safest solution rather than using a regex, but as you said, both methods would work.

**Comment by kondax** [comment]
I think it would be best to call the static URL, but without being hacky and modifying the returned parameters, we couldn't do that.

**Comment by scholtzm** [comment]
@kondax has already pointed this out kinda. The Laravel package is safe because it's always using `http://steamcommunity.com/openid` to verify the signature. Same with many other Steam specific packages.

However, `node-openid` will use query parameter (if provided) to verify the signature: https://github.com/havard/node-openid/blob/master/openid.js#L1243

`passport-steam` only receives `claimed_id` and this can be already spoofed any way you like.

**Comment by kondax** [comment]
@scholtzm if we bundled our own node-openid, we could fix the issue. I think this might be the only way to go forwards with it.

**Comment by mnzt** [comment]
@kondax I think that should be our very very last option

**Comment by kondax** [comment]
@mnzt Exactly. The only other option I see is creating a pull request on node-openid to allow specific validation URLs, or the same with node-passport.

**Comment by cocothewithlovein** [comment]
Oh god so this is what has been giving me nightmares for the past week, I've had rogue logins on my admin account which stole my sleep and was up for 3 days implementing multiple authentication checks, (express cid, jwt, ip checks, socket auth token) I'm glad the fault isn't in my code

**Comment by mnzt** [comment]
@cocothewithlovein You should never have admin access reliant on purely OAuth, you should always ensure 2FA on important applications.

We're currently stuck at a crossroad of keeping the dependency structure, or fixing underlying packages

**Comment by scholtzm** [comment]
@mnzt Since OpenID spec allows this ("supply any OpenID provider"), I don't think this will be changed in packages such as `node-openid`.

However, I think the fix has already been suggested - just validate the request itself.

A quick fix for anyone running this package right now would be to set `options.passReqToCallback` and then just validate `op_endpoint`, e.g.:

```js
// taken from the example
passport.use(new SteamStrategy({
    returnURL: 'http://localhost:3000/auth/steam/return',
    realm: 'http://localhost:3000/',
    apiKey: 'key',
    passReqToCallback: true
  },
  function(req, identifier, profile, done) {
    if(req.query['openid.op_endpoint'] !== 'https://steamcommunity.com/openid/login') {
      return done(null, false, { message: 'Claimed identity is invalid.' });
    }

    return done(null, profile);
  }
));
```

This can be done inside `passport-steam` as well.

**Comment by kondax** [comment]
@scholtzm That looks like the easiest solution.

**Comment by kondax** [comment]
I've created a pull request that adds @scholtzm's patch to the example. With the amount of people directly copying the example, it's probably considered better so that the deployed sites are patched from this exploit.

**Comment by Pepsai** [comment]
So the fix provided by @scholtzm would fix the exploit?

**Comment by kondax** [comment]
@Pepsai it should be a temporary fix, but yes, it would.

**Comment by cocothewithlovein** [comment]
@mnzt right, all the top level admin controls are done through a separate server so not any damages besides someone betting $3 through my account.

@kondax Thanks for that quick fix, now I can remove the 30 jwt's and ip checks that I had

**Comment by scholtzm** [comment]
@cocothewithlovein Can you add logging of `op_endpoints` if it's invalid? I'm curious to see what is being used.

I dug into this a bit further:

1. For whatever reason, `GetPlayerSummaries` WebAPI allows SteamIDs to be passed in in a very random format, e.g. this works `https://api.steampowered.com/ISteamUser/GetPlayerSummaries/v2/?key=KEY&format=json&steamids=http://steamcommunity.com.banana.com/openid/id/76561197960435530`.

2. This [line of code](https://github.com/liamcurry/passport-steam/blob/master/lib/passport-steam/strategy.js#L58) provides the whole identifier, but should only provide the SteamID itself.

3. Received identifier should be validated with RegEx (@kondax' previous PR was right about this, my bad 👍 ) to make sure it has the following format: `http://steamcommunity.com/openid/id/{7656\d+$}`

TL;DR Final patch should validate identifier format and ideally op_endpoint as well. I'm thinking this would have never happened in GetPlayerSummaries only accepted actual SteamIDs since then the identifier would have to be parsed in the first place.

**Comment by welps** [comment]
@scholtzm Appreciate all the help you've been contributing, just wanted to say thanks!

**Comment by kondax** [comment]
@scholtzm I thought my implementation was correct, you put me off haha. I'll update my latest PR to include the identifier checks and use that also.

**Comment by scholtzm** [comment]
@mnzt @welps This exploit should get more attention. 😃

**Comment by Pepsai** [comment]
@scholtzm  what do you mean by that?

**Comment by breatoz** [comment]
I know right; friggin scary

**Comment by welps** [comment]
@kondax @scholtzm @breatoz Sorry! Is #40 the fix I need to look at?

**Comment by kondax** [comment]
I'd suggest going with the fix mentioned above. No point with my commit as
there isn't a viable solution to fixing the issue.

Would be best to just have the check within the callback - we can't do
anything else.

**Comment by scholtzm** [comment]
@welps Yeah that's the one, but it's not done properly.

@kondax It can (and it should) be done within `passport-steam`. I can open new PR tomorrow if someone else doesn't do it first.

**Comment by kondax** [comment]
Feel free to do so now - but this has been inactive for a while and I think
the maintainers of the repo are just leaving it.

**Comment by welps** [comment]
@scholtzm Yeah, please open a PR, we'll review, and then merge it in. Thanks.

**Comment by kondax** [comment]
Sounds great.

**Comment by welps** [comment]
Fixed in #41 by @scholtzm 

Thanks everyone!
